<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment</b>
    
Hi, it's Sveta again. I was the reviewer on one of your previous projects, and I am going to review this one as well. 
    

Before we start, I want to pay your attention to the color marking:
    
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ✔️</b>
    
Great solutions and ideas that can and should be used in the future are in green comments.   
</div>    
    
    
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ⚠️</b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project.
</div>      
    
    
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment ❌</b>

Issues that need to be corrected to get right results are indicated in red comments. Note that the project cannot be accepted until these issues are resolved.
</div>    

<hr>
    
**Please, use some color other than those listed to highlight answers to my comments.**
I would also ask you **not to change, move or delete my comments** so that it would be easier for me to navigate during the next review.
    
In addition, my comments are defined as headings. 
They can mess up the content; however, they are convenient, since you can immediately go to them. I will remove the headings from my comments in the next review. 
   
    
    
<hr>
    
<font color='dodgerblue'>**A few words about the project:**</font> thank you so submitting the project! This is a hard topic, but you did a great job! Please use the correct path because I cannot run the code. I also skimmed the project and noticed several issues that need your attention. Please check them out! 
    
    
    

</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 2</b>
    

    
First, I see that my final comment about the project was not saved. Last time, I saved the notebook with the review, so I'll just copy my comment and paste it above. Sorry about that. Sometimes JupyterHub does not save the comments. 
    
    
Second, thank you for the updates. I left new comments with digit 2. Please take a look. I think they will help you finish the project :) I also expanded the comment about three datasets, so it should not confuse you. 

</div>


<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 3</b>


Great job, thank you very much for your diligence! My new comments are titled as **Reviewer's comment 3**. Would you take a look? :) 

    

    
</div>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 4</b>
    

Thanks a lot! New comments have digit 4. Please take a look.
    
I also want to note that there are some troubles on our platform and I had to add comments several times, since they were not saved. Please feel free to ask questions in case anything remains unclear. I will try to explain it and help you understand this difficult  topic.
</div>
<hr>
<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 5</b>
    
Just a couple of questions this time. 
    
</div>
<hr>
<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 6</b>
    
I have only one question this time. Please take a look.    
</div>
<hr>

<div style="border-radius: 15px; border: 3px solid indigo; padding: 15px;">
<b> Reviewer's comment 7</b>


Thank you so much for such a good job! This is a hard topic, but everything is great in your project! You can find my new comments with digit 7. I do not have any questions, so the project can be accepted. Good luck! 

    
</div>

## Prediction on Churning Customers 

#### This is a project to predict whether or not a customer will leave the bank soon. For this project i have the data on clients' past behavior and termination of contracts with the bank. I need an F1 score of at least .59

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
    
Please add a title and do not make the introduction a level 1 heading. 
    
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
    
Is it a title or intro? The introduction does not need to be level-1 title because it looks like a title. Please make it broader as well. You can add data description and outline the methods you are going to use here. 
</div>

<div class="alert alert-info" 
    
made a title and intro 
     
</div>

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment 3 </h2>
   

Try to not make your introductions smaller than the intermediate conclusions :) 
</div>

## Importing packages

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
import matplotlib.pyplot as plt
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score
from sklearn.utils import shuffle
from sklearn.utils import resample
import numpy as np
from sklearn.metrics import precision_recall_curve
from sklearn.metrics import roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
pd.options.mode.chained_assignment = None


## Opening dataset
### saving dataset as churn and preprocessing data

In [2]:
churn = pd.read_csv('/datasets/Churn.csv')

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    
Please use the correct path. 

</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Great, thank you so much! 
</div>

<div class="alert alert-info" 
    
fixed file path
     
</div>

In [3]:
churn

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0        2.0       0.00              1          1               1   
1        1.0   83807.86              1          0               1   
2        8.0  159660.80              3          1               0   
3        1.0       0.00              2          0               0   
4        2.0  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [4]:
churn.shape

(10000, 14)

In [5]:
churn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


#### checking for missing or duplicate values

In [6]:
churn_missing = churn.isnull().sum()
churn_duplicate = churn.duplicated().sum()

In [7]:
churn_missing

RowNumber            0
CustomerId           0
Surname              0
CreditScore          0
Geography            0
Gender               0
Age                  0
Tenure             909
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
Exited               0
dtype: int64

#### checking for duplicates

In [8]:
churn_duplicate

0

#### checking for duplicates in the customerid column

In [9]:
duplicate_customer = churn[churn.duplicated('CustomerId')]

In [10]:
duplicate_customer

Empty DataFrame
Columns: [RowNumber, CustomerId, Surname, CreditScore, Geography, Gender, Age, Tenure, Balance, NumOfProducts, HasCrCard, IsActiveMember, EstimatedSalary, Exited]
Index: []

#### no duplicates found in the customerid column

#### there are no duplicates that have been found

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Are there any duplicates in `CustomerId`?
    
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 3</b>
    
Are there any duplicates in `CustomerId` column?
    
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 4 </h2>
    
Correct!

</div>

<div class="alert alert-info" 
    
found no duplicates in the customerid column
     
</div>

#### there are no duplicates and 909 missing values in the tenure column

<div class="alert alert-info" 
     
checked for duplicates in all columns and found no duplicates
     
</div>

In [11]:
df_missing_tenure = churn[churn['Tenure'].isnull()]

In [12]:
df_missing_tenure.sample(10)

RowNumber  CustomerId      Surname  CreditScore Geography  Gender  Age  \
9041       9042    15724336        Yates          513   Germany  Female   49   
290         291    15652266   Chidiebele          703   Germany    Male   42   
9163       9164    15802368       Ch'eng          608    France  Female   47   
6170       6171    15611157     McElhone          709    France  Female   32   
8067       8068    15696834         Cone          530    France  Female   29   
981         982    15566156     Franklin          749   Germany  Female   44   
897         898    15782390        T'ien          621    France  Female   40   
6985       6986    15586699      Thomson          825    France    Male   32   
8614       8615    15809515        Lewis          797   Germany    Male   32   
5536       5537    15795878  Anayochukwu          636     Spain    Male   45   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9041     NaN  171601.27              1          1               0   
290      NaN   63227.00              1          0               1   
9163     NaN       0.00              1          1               1   
6170     NaN   87814.89              1          1               0   
8067     NaN       0.00              2          0               0   
981      NaN   71497.79              2          0               0   
897      NaN       0.00              1          1               0   
6985     NaN       0.00              2          0               0   
8614     NaN  151922.94              1          1               0   
5536     NaN       0.00              2          1               1   

      EstimatedSalary  Exited  
9041        126223.84       0  
290         137316.32       0  
9163        126012.57       0  
6170        138578.37       0  
8067        121451.21       0  
981         151083.80       0  
897         155155.25       0  
6985          9751.03       0  
8614          8877.06       0  
5536        159463.80       0

#### it looks like there is a wide range of age, credit score, and bank balance for the missing tenures. since there is no good explanation for why there are 909 missing values and they take up almost 10% of the data, i will fill them with the median tenure. 

In [13]:
median_value = churn['Tenure'].median()
churn['Tenure'].fillna(value=median_value, inplace=True)

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Median is usually a better choice.    
</div>

<div class="alert alert-info" 
     
changed mean to median
     
</div>

In [14]:
churn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


#### there are still object columns, machine learning wont work with objects. I need to either remove them or one hot encode them. 

In [15]:
object_columns = churn.select_dtypes(include=['object']).columns

In [16]:
object_columns

Index(['Surname', 'Geography', 'Gender'], dtype='object')

#### removing the Surname, RowNumber, and CustomerId column because it does not matter

In [17]:
churn_clean = churn.copy()
churn_clean.drop('Surname', axis=1, inplace=True)

<div class="alert alert-info" 

made a churn.copy() . thank you for the reminder
     
</div>

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Please, don't forget about the **copy** method. The problem is, if we don't use **copy** and change something in new dataframe, the old one may also have these changes. Example:    
</div>

In [18]:
# Reviewer's code 2.1

l = {'A':[1,2,4], 'B':[3,3,5]}

df1 = pd.DataFrame(l)

df2 = df1

df3 = df1.copy() 

display(df1, df2, df3)

A  B
0  1  3
1  2  3
2  4  5

A  B
0  1  3
1  2  3
2  4  5

A  B
0  1  3
1  2  3
2  4  5

In [19]:
# Reviewer's code 2.2

df2.loc[0, 'A'] = 'change2'

df3.loc[1, 'A'] = 'change3'

display(df1, df2, df3)

A  B
0  change2  3
1        2  3
2        4  5

A  B
0  change2  3
1        2  3
2        4  5

A  B
0        1  3
1  change3  3
2        4  5

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

As we see, the old dataframe, **df1** has the same changes as **df2** does, because **df2** was created without using **copy**. While the **df3** changes did not accur in the **df1** frame as **df3** was created with **copy**. </div>

In [20]:
churn_clean.drop(['CustomerId', 'RowNumber'], axis=1, inplace=True)

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    

You can drop a list of columns, which will make the code shorter. 
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3</h2>
    
Good. 

</div>

<div class="alert alert-info" 

fixed with one line of code
     
</div>

In [21]:
churn_clean

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619    France  Female   42     2.0       0.00              1   
1             608     Spain  Female   41     1.0   83807.86              1   
2             502    France  Female   42     8.0  159660.80              3   
3             699    France  Female   39     1.0       0.00              2   
4             850     Spain  Female   43     2.0  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9995          771    France    Male   39     5.0       0.00              2   
9996          516    France    Male   35    10.0   57369.61              1   
9997          709    France  Female   36     7.0       0.00              1   
9998          772   Germany    Male   42     3.0   75075.31              2   
9999          792    France  Female   28     5.0  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0             1               1        101348.88       1  
1             0               1        112542.58       0  
2             1               0        113931.57       1  
3             0               0         93826.63       0  
4             1               1         79084.10       0  
...         ...             ...              ...     ...  
9995          1               0         96270.64       0  
9996          1               1        101699.77       0  
9997          0               1         42085.58       1  
9998          1               0         92888.52       1  
9999          1               0         38190.78       0  

[10000 rows x 11 columns]

In [22]:
unique_geographies = churn_clean['Geography'].unique()
unique_geographies

array(['France', 'Spain', 'Germany'], dtype=object)

#### there are three unique geographies: France, Spain, and Germany

In [23]:
unique_gender = churn_clean['Gender'].unique()
unique_gender

array(['Female', 'Male'], dtype=object)

#### there are two unique genders: Female and Male

In [24]:
object_columns = churn_clean.select_dtypes(include=['object']).columns

In [25]:
object_columns

Index(['Geography', 'Gender'], dtype='object')

In [26]:
churn_clean_dummy = churn_clean.copy()

In [27]:
churn_clean_dummy = pd.get_dummies(churn_clean, columns=object_columns, drop_first=True)


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
Using `drop_first=True` may help you avoid multicollinearity.
    
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3</h2>
    
Correct. 
    
</div>

<div class="alert alert-info" 

added drop_first=True in the line of code
     
</div>

In [28]:
churn_clean_dummy

CreditScore  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             619   42     2.0       0.00              1          1   
1             608   41     1.0   83807.86              1          0   
2             502   42     8.0  159660.80              3          1   
3             699   39     1.0       0.00              2          0   
4             850   43     2.0  125510.82              1          1   
...           ...  ...     ...        ...            ...        ...   
9995          771   39     5.0       0.00              2          1   
9996          516   35    10.0   57369.61              1          1   
9997          709   36     7.0       0.00              1          0   
9998          772   42     3.0   75075.31              2          1   
9999          792   28     5.0  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  Geography_Germany  \
0                  1        101348.88       1                  0   
1                  1        112542.58       0                  0   
2                  0        113931.57       1                  0   
3                  0         93826.63       0                  0   
4                  1         79084.10       0                  0   
...              ...              ...     ...                ...   
9995               0         96270.64       0                  0   
9996               1        101699.77       0                  0   
9997               1         42085.58       1                  0   
9998               0         92888.52       1                  1   
9999               0         38190.78       0                  0   

      Geography_Spain  Gender_Male  
0                   0            0  
1                   1            0  
2                   0            0  
3                   0            0  
4                   1            0  
...               ...          ...  
9995                0            1  
9996                0            1  
9997                0            0  
9998                0            1  
9999                0            0  

[10000 rows x 12 columns]

#### creating a correlation_matrix to see if there is anything immediate standing out about exiting and any other variable

In [29]:
correlation_matrix = churn_clean_dummy.corr()


In [30]:
correlation_matrix

CreditScore       Age    Tenure   Balance  NumOfProducts  \
CreditScore           1.000000 -0.003965 -0.000060  0.006268       0.012238   
Age                  -0.003965  1.000000 -0.012606  0.028308      -0.030680   
Tenure               -0.000060 -0.012606  1.000000 -0.007535       0.011409   
Balance               0.006268  0.028308 -0.007535  1.000000      -0.304180   
NumOfProducts         0.012238 -0.030680  0.011409 -0.304180       1.000000   
HasCrCard            -0.005458 -0.011721  0.025979 -0.014858       0.003183   
IsActiveMember        0.025651  0.085472 -0.030681 -0.010084       0.009612   
EstimatedSalary      -0.001384 -0.007201  0.010049  0.012797       0.014204   
Exited               -0.027094  0.285323 -0.015989  0.118533      -0.047820   
Geography_Germany     0.005538  0.046897 -0.003153  0.401110      -0.010419   
Geography_Spain       0.004780 -0.001685  0.000772 -0.134892       0.009039   
Gender_Male          -0.002857 -0.027544  0.012041  0.012087      -0.021859   

                   HasCrCard  IsActiveMember  EstimatedSalary    Exited  \
CreditScore        -0.005458        0.025651        -0.001384 -0.027094   
Age                -0.011721        0.085472        -0.007201  0.285323   
Tenure              0.025979       -0.030681         0.010049 -0.015989   
Balance            -0.014858       -0.010084         0.012797  0.118533   
NumOfProducts       0.003183        0.009612         0.014204 -0.047820   
HasCrCard           1.000000       -0.011866        -0.009933 -0.007138   
IsActiveMember     -0.011866        1.000000        -0.011421 -0.156128   
EstimatedSalary    -0.009933       -0.011421         1.000000  0.012097   
Exited             -0.007138       -0.156128         0.012097  1.000000   
Geography_Germany   0.010577       -0.020486         0.010297  0.173488   
Geography_Spain    -0.013480        0.016732        -0.006482 -0.052667   
Gender_Male         0.005766        0.022544        -0.008112 -0.106512   

                   Geography_Germany  Geography_Spain  Gender_Male  
CreditScore                 0.005538         0.004780    -0.002857  
Age                         0.046897        -0.001685    -0.027544  
Tenure                     -0.003153         0.000772     0.012041  
Balance                     0.401110        -0.134892     0.012087  
NumOfProducts              -0.010419         0.009039    -0.021859  
HasCrCard                   0.010577        -0.013480     0.005766  
IsActiveMember             -0.020486         0.016732     0.022544  
EstimatedSalary             0.010297        -0.006482    -0.008112  
Exited                      0.173488        -0.052667    -0.106512  
Geography_Germany           1.000000        -0.332084    -0.024628  
Geography_Spain            -0.332084         1.000000     0.016889  
Gender_Male                -0.024628         0.016889     1.000000

#### nothing immediate jumps out as highly correlative between exiting and anything else

#### getting an overview of the dataset to see where the tendencies are

In [31]:
descriptive_stats = churn_clean_dummy.describe()




In [32]:
descriptive_stats

CreditScore           Age       Tenure        Balance  NumOfProducts  \
count  10000.000000  10000.000000  10000.00000   10000.000000   10000.000000   
mean     650.528800     38.921800      4.99790   76485.889288       1.530200   
std       96.653299     10.487806      2.76001   62397.405202       0.581654   
min      350.000000     18.000000      0.00000       0.000000       1.000000   
25%      584.000000     32.000000      3.00000       0.000000       1.000000   
50%      652.000000     37.000000      5.00000   97198.540000       1.000000   
75%      718.000000     44.000000      7.00000  127644.240000       2.000000   
max      850.000000     92.000000     10.00000  250898.090000       4.000000   

         HasCrCard  IsActiveMember  EstimatedSalary        Exited  \
count  10000.00000    10000.000000     10000.000000  10000.000000   
mean       0.70550        0.515100    100090.239881      0.203700   
std        0.45584        0.499797     57510.492818      0.402769   
min        0.00000        0.000000        11.580000      0.000000   
25%        0.00000        0.000000     51002.110000      0.000000   
50%        1.00000        1.000000    100193.915000      0.000000   
75%        1.00000        1.000000    149388.247500      0.000000   
max        1.00000        1.000000    199992.480000      1.000000   

       Geography_Germany  Geography_Spain   Gender_Male  
count       10000.000000     10000.000000  10000.000000  
mean            0.250900         0.247700      0.545700  
std             0.433553         0.431698      0.497932  
min             0.000000         0.000000      0.000000  
25%             0.000000         0.000000      0.000000  
50%             0.000000         0.000000      1.000000  
75%             1.000000         0.000000      1.000000  
max             1.000000         1.000000      1.000000

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
    
> creating several datasets based on the countr
    
    
For what purpose? This is a feature that you can take into account in training. If you want to split the dataframe based on the features, then we can create 100 dataframes here based on genre, region, age and so on. But we do not do that, right? :) 
    
</div>

<div class="alert alert-info" 

sorry, I thought you meant I had to create three dataframes. I know now that what you meant was I need to create 
3 datasets. the train, valid, and test sets. I will fix this
     
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3</h2>
    
That's what I thought. No worries! :) 
</div>

#### creating target and features for my training, validation, and test sets

In [33]:
target = churn_clean_dummy['Exited']
features = churn_clean_dummy.drop('Exited', axis=1)

features_train_valid, features_test, target_train_valid, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345, stratify=churn_clean_dummy['Exited'])


features_train, features_valid, target_train, target_valid = train_test_split(
    features_train_valid, target_train_valid, test_size=0.25, random_state=12345, stratify=target_train_valid)


In [34]:
features_train_valid.shape

(7500, 11)

In [35]:
features_test.shape

(2500, 11)

In [36]:
target_train_valid.shape

(7500,)

In [37]:
target_test.shape

(2500,)

In [38]:
features_train.shape

(5625, 11)

In [39]:
features_valid.shape

(1875, 11)

In [40]:
target_train.shape

(5625,)

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 3</b>
    
    
Correct, but we need three sets: train, valid, and test. There's no test set. Would you update the cell above? 
    
</div>

</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 4 </h2>
 
Correct! 
    
</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 4 </b>
    

However, it may be useful to print their sizes. 
</div>
<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 5 </h2>
    
Good.
    
</div>

<div class="alert alert-info" 

updated cell above with a test set     
     
</div>

#### scaling the creditscore, age, estimatedsalary, balance, tenure, and numofproducts columns

In [41]:
scaler = StandardScaler()
numeric = ['CreditScore', 'Age', 'EstimatedSalary', 'Balance', 'Tenure', 'NumOfProducts']

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 5</b>
   

Scaler was created twice :)     
</div>

#### fitting the features training set and valid set with the scaled numbers

In [42]:
features_train[numeric] = scaler.fit_transform(features_train[numeric])

features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

features_train.head()

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
2461    -0.142911  0.105354  1.092507  0.776076      -0.917355          1   
1995    -0.688408  0.492748  0.002197  0.298591       0.796402          1   
2926    -1.450045  2.332872  1.092507 -1.234405       0.796402          1   
2902     0.412878 -0.572586 -1.451550  0.821519      -0.917355          1   
7150     0.495217 -0.669435 -0.361240  0.537144      -0.917355          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
2461               1         1.286704                  0                0   
1995               1         0.101459                  0                0   
2926               1         0.253635                  0                0   
2902               1         0.223798                  0                0   
7150               1         0.465285                  0                0   

      Gender_Male  
2461            1  
1995            0  
2926            1  
2902            0  
7150            0

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
Please explain why are creating so many datasets and remove specific features. Did you conduct experiments? If yes, provide the results please to make it clear. Otherwise, a reader may not understand why you chose these features.  
    
</div>

<div class="alert alert-info" 

I explained above why I created so many datasets. fixing now
     
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 3</h2>
    
Yes, my comment above is not relevant anymore. 
</div>

### decision tree model (unbalanced)

In [43]:
decision_tree = DecisionTreeClassifier(random_state=12345)
decision_tree.fit(features_train, target_train)
dt_predictions = decision_tree.predict(features_valid)
dt_f1_score = f1_score(target_valid, dt_predictions)
print(f"Decision Tree F1 score: {dt_f1_score: .2f}")


Decision Tree F1 score:  0.49


### decision tree model (balanced)

In [44]:
decision_tree = DecisionTreeClassifier(random_state=12345, class_weight='balanced')
decision_tree.fit(features_train, target_train)
dt_predictions = decision_tree.predict(features_valid)
dt_f1_score = f1_score(target_valid, dt_predictions)
print(f"Decision Tree F1 score: {dt_f1_score: .2f}")


Decision Tree F1 score:  0.48


#### unbalanced dt has a higher f1 score

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 3</b>
    

Since our main goal is to choose a mode with the highest F` score, let's print it here and further. Keep in mind that F1 score and accuracy are not the same metrics. Accuracy does not take into account data distribution, so it is better to use F1 when the distribution is not balanced. For this reason, try to choose a model based on F1 score, not accurancy. 
</div>

<div class="alert alert-info" 

I see. I thought I was supposed to choose which one had the highest accuracy and then build the f1 score from the one with the highest accuracy. will update by taking the f1 score instead of accuracy
     
</div>

### Random forest model (unbalanced)

In [45]:
random_forest = RandomForestClassifier(random_state=12345)
random_forest.fit(features_train, target_train)
rf_predictions = random_forest.predict(features_valid)
rf_f1_score = f1_score(target_valid, rf_predictions)
print(f"Random Forest F1 score: {rf_f1_score:.2f}")

Random Forest F1 score: 0.55


### Random forest model (balanced)

In [46]:
random_forest = RandomForestClassifier(random_state=12345, class_weight='balanced')
random_forest.fit(features_train, target_train)
rf_predictions = random_forest.predict(features_valid)
rf_f1_score = f1_score(target_valid, rf_predictions)
print(f"Random Forest F1 score: {rf_f1_score:.2f}")

Random Forest F1 score: 0.53


#### random forest model unbalanced has a higher f1 score than all the models so far

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 4 </h2>

Good.    
</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment 4 </h2>
 
It may reasonable, however, to try different hyperparameters.
</div>

### logistic regression (unbalanced)

In [47]:
logistic_regression = LogisticRegression(max_iter=1000, random_state=12345)
logistic_regression.fit(features_train, target_train)
lr_predictions = logistic_regression.predict(features_valid)
lr_f1_score = f1_score(target_valid, lr_predictions)
print(f"Logistic Regression F1 score: {lr_f1_score:.2f}")

Logistic Regression F1 score: 0.31


### logistic regression(balanced)

In [48]:
logistic_regression = LogisticRegression(max_iter=1000, random_state=12345, class_weight='balanced')
logistic_regression.fit(features_train, target_train)
lr_predictions = logistic_regression.predict(features_valid)
lr_f1_score = f1_score(target_valid, lr_predictions)
print(f"Logistic Regression F1 score: {lr_f1_score:.2f}")

Logistic Regression F1 score: 0.47


#### logistic regression balanced f1 score is higher than the unbalanced score but is still lower than the random forest unbalanced model

### downsampling the majority and testing models again to try and achieve a higher f1 score

In [49]:
majority_class = target_train.value_counts().idxmax()
minority_class = target_train.value_counts().idxmin()

majority_class_indices = features_train[target_train == majority_class]
minority_class_indices = features_train[target_train == minority_class]

majority_downsampled = resample(majority_class_indices,
                                replace=False,  
                                n_samples=len(minority_class_indices),  
                                random_state=12345)  


features_downsampled = pd.concat([majority_downsampled, minority_class_indices])
target_downsampled = pd.concat([target_train[target_train == majority_class].loc[majority_downsampled.index],
                                target_train[target_train == minority_class]])


<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 4 </b>
    

Shouldn't we apply resampling to train data only?
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 5 </b>
    

You do not need to apply `train_test_split` again, since these samples are created. Moreover, why are you using target there? You already have a traning set created in the previous section. Just use it here. 


The resampling part is indeed quite hard to comprehend sometimes. Please ask don't hesitate to ask our tutors for help if you need that. 

</div>

<div class="alert alert-info" 

applied resampling to train data only
     
</div>

#### decision tree with downsampled majority

In [50]:
decision_tree = DecisionTreeClassifier(random_state=12345)
decision_tree.fit(features_downsampled, target_downsampled)
dt_predictions = decision_tree.predict(features_valid)
dt_f1_score = f1_score(target_valid, dt_predictions)
print(f"Decision Tree F1 score: {dt_f1_score:.2f}")

Decision Tree F1 score: 0.46


#### random forest with downsampled majority

In [51]:
random_forest = RandomForestClassifier(random_state=12345)
random_forest.fit(features_downsampled, target_downsampled)
rf_predictions = random_forest.predict(features_valid)
rf_f1_score = f1_score(target_valid, rf_predictions)
print(f"Random Forest F1 score: {rf_f1_score:.2f}")

Random Forest F1 score: 0.56


#### logistic regression with downsampled majority

In [52]:
logistic_regression = LogisticRegression(max_iter=1000, random_state=12345, solver='liblinear')
logistic_regression.fit(features_downsampled, target_downsampled)
lr_predictions = logistic_regression.predict(features_valid)
lr_f1_score = f1_score(target_valid, lr_predictions)
print(f"Logistic Regression F1 score: {lr_f1_score:.2f}")

Logistic Regression F1 score: 0.48


#### random forest with downsampled majority scored the highest so far with an f1 score of .75

### Upsampling the minority to try and achieve a higher f1 score

In [53]:
minority_upsampled = resample(minority_class_indices,
                              replace=True,  
                              n_samples=len(majority_class_indices),  
                              random_state=12345)  

features_upsampled = pd.concat([majority_class_indices, minority_upsampled])
target_upsampled = pd.concat([target_train[target_train == majority_class],
                              target_train[target_train == minority_class].loc[minority_upsampled.index]])

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 5 </b>
    
Same issue. 

</div>

#### decision tree with upsampled minority

In [54]:
decision_tree = DecisionTreeClassifier(random_state=12345)
decision_tree.fit(features_upsampled, target_upsampled)
dt_predictions = decision_tree.predict(features_valid)
dt_f1_score = f1_score(target_valid, dt_predictions)
print(f"Decision Tree F1 score: {dt_f1_score:.2f}")

Decision Tree F1 score: 0.43


#### random forest with upsampled minority

In [55]:
random_forest = RandomForestClassifier(random_state=12345)
random_forest.fit(features_upsampled, target_upsampled)
rf_predictions = random_forest.predict(features_valid)
rf_f1_score = f1_score(target_valid, rf_predictions)
print(f"Random Forest F1 score: {rf_f1_score:.2f}")

Random Forest F1 score: 0.58


#### logistic regression with upsampled minority

In [56]:
logistic_regression = LogisticRegression(max_iter=1000, random_state=12345, solver='liblinear')
logistic_regression.fit(features_upsampled, target_upsampled)
lr_predictions = logistic_regression.predict(features_valid)
lr_f1_score = f1_score(target_valid, lr_predictions)
print(f"Logistic Regression F1 score: {lr_f1_score:.2f}")

Logistic Regression F1 score: 0.47


#### random forest with upsampled minority scored the highest with an f1 score of .58. I will try and find the best hyperparameters for this model

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment 6 </h2>
   

You should do that each time you check the model. For instance, some group of parameters may work better with the upsampled data than we the downsampled data. Therefore, it is always a good idea to try different parameters on each step you took above.     
</div>

In [57]:
best_f1_score = 0
best_est = 0


for est in range(1, 55):  
    rf = RandomForestClassifier(random_state=12345, n_estimators=est)
    rf.fit(features_upsampled, target_upsampled)
    predictions_valid = rf.predict(features_valid)
    f1 = f1_score(target_valid, predictions_valid)  
    if f1 > best_f1_score:
        best_f1_score = f1  
        best_est = est  


print("Best F1 score on the validation set (n_estimators = {}): {:.2f}".format(best_est, best_f1_score))


final = RandomForestClassifier(random_state=12345, n_estimators=best_est)
final.fit(features_upsampled, target_upsampled)

Best F1 score on the validation set (n_estimators = 44): 0.59


RandomForestClassifier(n_estimators=44, random_state=12345)

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 7 </h2>
    
Great!     
    
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 3</b>
    

Let me just repeat the comments that were moved to the very end of the project. 
    
    


    - Before we resample the data, let's try to use `class_weight` parameter for each of the models you use above. 
    

    - Compare all models with this parameter. It is possible that regression will have a better performance, so we need to compare all three models with unbalanced data, with the `class_weight = 'balanced'`, and with the re-sampling. Then we will be able to choose the best model with the best setting. After you choose a model, train the data on full set (train + valid) and use test set to calculate the F1 score. 



Let's summarize. We need 1. Compare basic models; 2. Compare the same models with the updated `class_weight` parameter; 3. Compare the same models with the resampling methods (here I recommend that you try both upsampling and downsampling). 

Then we analyze the results a choose the best one. We then use a chosen model, be it regression with the upsampling or a simpel tree, train the data again on train+valid sets (combined) and test the model on test data. 

</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 4 </h2>
    
Good job!

</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 4</b>
    
However,the results may change. In addition, try to train a chosen model with chosen settings on full data (train set combined with validation set).</div>

<div class="alert alert-info" 

model with full data
     
</div>

In [58]:
combined_features = pd.concat([features_train, features_valid])
combined_target = pd.concat([target_train, target_valid])


In [59]:
majority_class = combined_target.mode()[0]
minority_class = 1 if majority_class == 0 else 0


majority_features = combined_features[combined_target == majority_class]
minority_features = combined_features[combined_target == minority_class]
majority_target = combined_target[combined_target == majority_class]
minority_target = combined_target[combined_target == minority_class]


minority_features_upsampled, minority_target_upsampled = resample(minority_features, minority_target,
                                                                  replace=True,
                                                                  n_samples=len(majority_features),
                                                                  random_state=54321)


upsampled_features = pd.concat([majority_features, minority_features_upsampled])
upsampled_target = pd.concat([majority_target, minority_target_upsampled])


model = RandomForestClassifier(random_state=12345, n_estimators=100)
model.fit(upsampled_features, upsampled_target)

predictions = model.predict(features_test)
f1 = f1_score(target_test, predictions)

print(f"F1 score on the upsampled combined dataset: {f1:.2f}")

F1 score on the upsampled combined dataset: 0.59


#### The final f1 score on the upsampled combined dataset with the best model is .59

In [60]:
probabilities = model.predict_proba(features_test)[:, 1] 

auc_roc_final = roc_auc_score(target_test, probabilities)
print(f"auc_roc score on the upsampled combined dataset: {auc_roc_final:.2f}")

auc_roc score on the upsampled combined dataset: 0.86


#### the final auc_roc score is .86

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 6 </b>
    

Almost 100%. Isn't it strange? Don't we have a leakage? 

</div>

<div class="alert alert-info" 

fixed with testing the model on features test and target test with the combined features train, target train, features valid, target valid
     
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 7 </h2>
    
Yes, everything is correct now.      
    
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b>  Reviewer's comment 5 </b>
    

If the best result was received with resampling, shouldn't we apply it here as well? 
    
    
To make it simplier, try to create the resampling function. 

</div>

## Conclusion
#### in conclusion of this project. I had the churn data for a bank with 10000 customers as a sample base. I first cleaned the data, dropped what I did not need, scaled it, then tried several different models to achieve a high accuracy score. I had the best accuracy score and f1 score with the random forest model. I then balanced the data by upscaling the minority because of the high class imbalance between those who have and have not exited the company. After balancing and checking I achieved a higher f1 score but did not meet the required 59% f1 score yet. I upsampled the minority which gave me the highest f1 score. After testing the model on the test set I achieved an f1 score of 59%  and an auc-roc score of 86%. This model will serve well for the banks purposes.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 7 </h2>
    
Very good!     
</div>

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 3</b>
    

Next time, please, do not move reviewer's comments. This is exactly what I asked for in the very first comment. It is very hard to remember all the issues a project had without the comments where I left them. It is also quite hard to read them here and then look for a specific location in the project where I left them last time.  
</div>

<div class="alert alert-info" 

sorry, things got messy between deleting cells and adding new ones. didn't mean to move all the comments
     
</div>

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    

    test = churn_clean
    test_clean = test
    
    
Isn't it just `test_clean = churn_clean`? 
    
    
Please do not forget to use the `copy` method if you do not want to accidentally change anything in the initial dataframe. 
    
    
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
Reconsider the way you split the data please. Keep in mind that since we are supposed to receive high F1 score on test data, we need 3 datasets. 

</div>
<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2>  Reviewer's comment </h2>
   

After you successfully split the data, you can scale it, it never hurts.

</div>

<div class="alert alert-info" 
    
created four datasets based on the countries and one without countries and gender. data has been scaled
     
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
> created four datasets based on the countries and one without countries and gender. data has been scaled

    
Yes, but the reasoning behind that is unclear. You do not need to split data this way. 
    
    
 
Don't forget that we need 3 datasets (train, valid, and test). So, take the initial dataset, apply encoding, split it into 3 parts, and scale. 
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Please compare several models here and choose the best one on validation data. Test data has to be used to get the final F1 score on a chosen model. 
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
For each model you use, print F1 score please. 



> I will use the Decision Tree Model for my machine learning


Why? Elaborate on this please, since you did not say anything about Tree and Forest models. 

</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment</h2>
    
Are the classes balanced? 
</div>

<div class="alert alert-info" 
    
balanced the classes 
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<h2> Reviewer's comment 2</h2>
    
Correct, there's a noticeable imbalance.     
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
> for the france dataset


So, we chose the best model for France, while our goal is be able to get a prediction for people from different countries :) That is the reason why we do not need to create dataframes with one feature value in this task :) 

</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment</b>
    

Before we resample the data, let's try to use `class_weight` parameter for each of the models you use above. 
    
</div>
<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
Compare all models with this parameter. It is possible that regression will have a better performance, so we need to compare all three models with unbalanced data, with the `class_weight = 'balanced'`, and with the re-sampling. Then we will be able to choose the best model with the best setting. After you choose a model, train the data on full set (train + valid) and use test set to calculate the F1 score. 
</div>

<div class="alert alert-info" 
    
used the class_weight balanced parameter for the lr model
</div>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">
<b> Reviewer's comment 2</b>
    
Please make the conclusion broader: what we had, what results we received, and what van be inferred from these results. 
</div>